### PR TITLE
Add support for OTEL_SERVICE_NAME environmental variable

### PR DESCRIPTION
**Pull request by pellared**
Partially addresses #1453.

This feature is important so that we can reuse the SDK in https://github.com/open-telemetry/opentelemetry-dotnet-instrumentation.

## Changes

- `ResourceBuilder.AddEnvironmentVariableDetector` handles `OTEL_SERVICE_NAME`
   environmental variable.

### CODE REVIEW

File: src/OpenTelemetry/Resources/ResourceBuilderExtensions.cs
Comment (pellared) [inline src/OpenTelemetry/Resources/ResourceBuilderExtensions.cs:120]: Couldn't this detector be out-of-the-box? This is what other languages do.

_For sure something for a separate PR and maybe even needs some GH Issue for discussion._

File: src/OpenTelemetry/README.md
Comment (reyang) [inline src/OpenTelemetry/README.md:20]: Both `Configuration Parameters` and `Remarks` seem to be too detailed and generic for the ToC. Perhaps we want to remove the H3 (e.g. replace `### Remarks` with `**Remarks:**`).

File: src/OpenTelemetry/README.md
Comment (reyang) [inline src/OpenTelemetry/README.md:253]: nit:

```suggestion
| -------------------------- | -------------------------------------------------- |
```

File: src/OpenTelemetry/Resources/ResourceBuilderExtensions.cs
Comment (reyang) [reply src/OpenTelemetry/Resources/ResourceBuilderExtensions.cs:120]: +1 on making this out-of-the-box

**Comment by pellared** [comment]
Flaky test:

```
Failed OpenTelemetry.Instrumentation.AspNetCore.Tests.MetricTests.RequestMetricIsCaptured [113 ms]
  Error Message:
   System.Exception : Unsupported Type
  Stack Trace:
     at OpenTelemetry.Metrics.SumMetricAggregator.get_Sum() in /home/runner/work/opentelemetry-dotnet/opentelemetry-dotnet/src/OpenTelemetry/Metrics/MetricAggregators/SumMetricAggregator.cs:line 74
   at OpenTelemetry.Instrumentation.AspNetCore.Tests.MetricTests.RequestMetricIsCaptured() in /home/runner/work/opentelemetry-dotnet/opentelemetry-dotnet/test/OpenTelemetry.Instrumentation.AspNetCore.Tests/MetricTests.cs:line 93
--- End of stack trace from previous location ---
fail: Microsoft.AspNetCore.Diagnostics.DeveloperExceptionPageMiddleware[1]
      An unhandled exception has occurred while executing the request.
      System.Exception: exception description
         at OpenTelemetry.Instrumentation.AspNetCore.Tests.IncomingRequestsCollectionsIsAccordingToTheSpecTests.TestCallbackMiddlewareImpl.ProcessAsync(HttpContext context) in /home/runner/work/opentelemetry-dotnet/opentelemetry-dotnet/test/OpenTelemetry.Instrumentation.AspNetCore.Tests/IncomingRequestsCollectionsIsAccordingToTheSpecTests.cs:line 176
         at TestApp.AspNetCore._5._0.CallbackMiddleware.InvokeAsync(HttpContext context) in /home/runner/work/opentelemetry-dotnet/opentelemetry-dotnet/test/TestApp.AspNetCore.5.0/CallbackMiddleware.cs:line 35
         at Microsoft.AspNetCore.Diagnostics.DeveloperExceptionPageMiddleware.Invoke(HttpContext context)
warn: Microsoft.AspNetCore.Diagnostics.DeveloperExceptionPageMiddleware[2]
      The response has already started, the error page middleware will not be executed.
fail: Microsoft.AspNetCore.Diagnostics.DeveloperExceptionPageMiddleware[1]
      An unhandled exception has occurred while executing the request.
      System.Exception: exception description
         at OpenTelemetry.Instrumentation.AspNetCore.Tests.IncomingRequestsCollectionsIsAccordingToTheSpecTests.TestCallbackMiddlewareImpl.ProcessAsync(HttpContext context) in /home/runner/work/opentelemetry-dotnet/opentelemetry-dotnet/test/OpenTelemetry.Instrumentation.AspNetCore.Tests/IncomingRequestsCollectionsIsAccordingToTheSpecTests.cs:line 176
         at TestApp.AspNetCore._5._0.CallbackMiddleware.InvokeAsync(HttpContext context) in /home/runner/work/opentelemetry-dotnet/opentelemetry-dotnet/test/TestApp.AspNetCore.5.0/CallbackMiddleware.cs:line 35
         at Microsoft.AspNetCore.Diagnostics.DeveloperExceptionPageMiddleware.Invoke(HttpContext context)
warn: Microsoft.AspNetCore.Diagnostics.DeveloperExceptionPageMiddleware[2]
      The response has already started, the error page middleware will not be executed.
```

**Review by dszmigielski — APPROVED** [review]

File: src/OpenTelemetry/README.md
Comment (pjanotti) [inline src/OpenTelemetry/README.md:248]:
```suggestion
It is also possible to configure the `Resource` by using `AddEnvironmentVariableDetector`
```

File: src/OpenTelemetry/README.md
Comment (pjanotti) [inline src/OpenTelemetry/README.md:254]: Perhaps link add a link to the spec (permalink to the version currently implemented).

File: src/OpenTelemetry/Resources/OtelServiceNameEnvVarDetector.cs
Comment (pjanotti) [inline src/OpenTelemetry/Resources/OtelServiceNameEnvVarDetector.cs:31]: nit:
```suggestion
            if (!string.IsNullOrWhiteSpace(envResourceAttributeValue))
```

Comment (pellared) [reply src/OpenTelemetry/Resources/OtelServiceNameEnvVarDetector.cs:31]: `IsNullOrEmpty` is used a lot more frequently than `IsNullOrWhiteSpace`.
I would prefer to stick with `IsNullOrEmpty` for sake of consistency.

File: src/OpenTelemetry/README.md
Comment (pellared) [reply src/OpenTelemetry/README.md:20]: @open-telemetry/dotnet-maintainers WDYT - should I change it?

BTW I think that the `Propagators` section should be under `Tracing configuration` - and not under `Advanced topics`.

File: src/OpenTelemetry/README.md
Comment (pellared) [reply src/OpenTelemetry/README.md]: https://github.com/open-telemetry/opentelemetry-dotnet/pull/2209/commits/d5198652166c11a4ae5cfb893b466094eba8dffe

File: src/OpenTelemetry/Resources/OtelServiceNameEnvVarDetector.cs
Comment (pjanotti) [reply src/OpenTelemetry/Resources/OtelServiceNameEnvVarDetector.cs:31]: Good point, since it is the pattern on the code base then let's stick to it.

**Review by pjanotti — APPROVED** [review]

File: src/OpenTelemetry/Resources/ResourceBuilderExtensions.cs
Comment (pellared) [reply src/OpenTelemetry/Resources/ResourceBuilderExtensions.cs:120]: I will make a separate PR to address it

File: src/OpenTelemetry/Resources/OtelServiceNameEnvVarDetector.cs
Comment (CodeBlanch) [inline src/OpenTelemetry/Resources/OtelServiceNameEnvVarDetector.cs:24]: Would you mind making this `public const string` (class is already internal) and referencing it in the unit tests so we don't need a duplicate value there? I think this assembly already has internals visible to the tests.

File: src/OpenTelemetry/Resources/OtelServiceNameEnvVarDetector.cs
Comment (CodeBlanch) [inline src/OpenTelemetry/Resources/OtelServiceNameEnvVarDetector.cs:30]: Might want to mirror this `SecurityException` logic here:

https://github.com/open-telemetry/opentelemetry-dotnet/blob/54d46d80a03acfa0813021e5875a00a7ef423274/src/OpenTelemetry.Exporter.Jaeger/JaegerExporterOptions.cs#L54

Comment (pellared) [reply src/OpenTelemetry/Resources/OtelServiceNameEnvVarDetector.cs:30]: Will do

File: src/OpenTelemetry/Resources/OtelServiceNameEnvVarDetector.cs
Comment (pellared) [reply src/OpenTelemetry/Resources/OtelServiceNameEnvVarDetector.cs]: Addressed https://github.com/open-telemetry/opentelemetry-dotnet/pull/2209/commits/3268d715d580846c60e3d72eed99e09f8dd073c9

File: src/OpenTelemetry/Resources/OtelServiceNameEnvVarDetector.cs
Comment (pellared) [reply src/OpenTelemetry/Resources/OtelServiceNameEnvVarDetector.cs]: Addressed e0a9aa40

File: src/OpenTelemetry/Resources/OtelEnvResourceDetector.cs
Comment (CodeBlanch) [inline src/OpenTelemetry/Resources/OtelEnvResourceDetector.cs:44]: Should we switch this catch to `SecurityException` to match the others?

**Review by CodeBlanch — APPROVED** [review]
LGTM

File: src/OpenTelemetry/Resources/OtelEnvResourceDetector.cs
Comment (pellared) [reply src/OpenTelemetry/Resources/OtelEnvResourceDetector.cs:44]: I am not sure if this is the only exception type that can happen here.

**Comment by pellared** [comment]
Any idea why EasyCLA is not being triggered?

**Comment by cijothomas** [comment]
> Any idea why EasyCLA is not being triggered?

no! same issue with another PR also

File: src/OpenTelemetry/Resources/OtelEnvResourceDetector.cs
Comment (CodeBlanch) [reply src/OpenTelemetry/Resources/OtelEnvResourceDetector.cs:44]: I think it OK to just catch `SecurityException` but defer to @cijothomas. I'm basically just echoing stuff that was discussed on this PR: https://github.com/open-telemetry/opentelemetry-dotnet/pull/2123#discussion_r667134036

**Comment by pellared** [comment]
@cijothomas EasyCLA is working

File: src/OpenTelemetry/Resources/OtelEnvResourceDetector.cs
Comment (pellared) [reply src/OpenTelemetry/Resources/OtelEnvResourceDetector.cs:44]: I tested it as much as possible and I think it is OK to just catch `SecurityException`

File: src/OpenTelemetry/README.md
Comment (pellared) [reply src/OpenTelemetry/README.md:20]: I will make a separate PR for it as I did not get feedback on what is the preference

**Comment by pellared** [comment]
@open-telemetry/dotnet-maintainers PTAL

**Review by cijothomas — APPROVED** [review]

**Pull request merged**
